### ISSUE TITLE
Enable Adapter is not working exactly the same as when the adapter is merged to the Codellama Model

### ISSUE THREAD

**Issue by akhilamartia** [huggingface/peft#1226]
### System Info

the codellama model from hugging face :(https://huggingface.co/codellama/CodeLlama-7b-Instruct-hf) is Peft Lora tuned to have an adapter.

So i have 2 versions, one where i add the adapter to the model dynamically, enable/set it and test it with a sample code completion

```python
BASE_MODEL_ID = 'codellama/CodeLlama-7b-Instruct-hf'
from transformers import AutoModelForCausalLM, AutoTokenizer
model = AutoModelForCausalLM.from_pretrained(BASE_MODEL_ID)
from peft import PeftConfig, PeftModel
peft_config = PeftConfig.from_pretrained(ADAPTER_ID) #custom adapter
peft_config.init_lora_weights = False
model.add_adapter(peft_config, adapter_name="code_adapter")
```
the second version is where i merged the pretrained model with adapter and tested with sample code completion request.

and the output is different in both cases even when there is no change in the input or output tokens, everything is the same.

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [X] My own task or dataset (give details below)

### Reproduction

```python
from peft import PeftConfig, PeftModel
peft_config = PeftConfig.from_pretrained(ADAPTER_ID)
# to initiate with random weights
peft_config.init_lora_weights = False
model.add_adapter(peft_config, adapter_name="code_adapter")
prompt = "write java code for fetching appName using the vastid"
tokenizer = AutoTokenizer.from_pretrained(BASE_MODEL_ID)
dummy_input = tokenizer(text, return_tensors="pt")
peft_logits = model.generate(**dummy_input)

#Merged Model Code:
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
from peft import PeftModel, PeftConfig, LoraConfig, TaskType
import time
import json
if merge:
    pretrained_model_path = "codellama/CodeLlama-7b-Instruct-hf"
    model = AutoModelForCausalLM.from_pretrained(pretrained_model_path)
    peft_id_Equinox = "codellama_adapters"
    model = PeftModel.from_pretrained(model, peft_id_Equinox, adapter_name= "codellama_merged_lora
    model = model.merge_and_unload()
    model.save_pretrained("codellama_merged_lora")
```

### Expected behavior

The output of enable adapter model and merged model has to be same

**Comment by younesbelkada** [comment]
To add what @BenjaminBossan we also have strong CI tests that makes sure the logits are the same when the adapter is merged. Make also sure to always use the greedy generation strategy `do_sample=False` when calling `generate`. A full reproducer would be really helpful !

**Comment by akhilamartia** [comment]
> A few points:
> 
> * Did you fix the random seed for the two attempts?
> * Are the results wildly different or just slightly different (merging will not result in 100% the same outputs)
> * You use `peft_config.init_lora_weights = False` when loading the pretrained adapter, which you shouldn't do (but probably that's not the issue)
> * You don't show the merging code, so I can't judge if there is anything wrong there

I have updated the issue with the merge code now.

**Comment by BenjaminBossan** [comment]
Thanks for updating the code. This still doesn't answer the questions above. Here are a few things to try:

- please try to fix the random seed
- ensure that the model is in `eval` mode
- check the logits, not just the generations -- are they similar or different?
- if you compare the logits immediately before and immediately after merging, is there a big difference?

**Comment by akhilamartia** [comment]
> A few points:
> 
> * Did you fix the random seed for the two attempts?
> * Are the results wildly different or just slightly different (merging will not result in 100% the same outputs)
> * You use `peft_config.init_lora_weights = False` when loading the pretrained adapter, which you shouldn't do (but probably that's not the issue)
> * You don't show the merging code, so I can't judge if there is anything wrong there



> Thanks for updating the code. This still doesn't answer the questions above. Here are a few things to try:
> 
> * please try to fix the random seed
> * ensure that the model is in `eval` mode
> * check the logits, not just the generations -- are they similar or different?
> * if you compare the logits immediately before and immediately after merging, is there a big difference?

1. yes, i fixed the random seed issue. and do_sample is set to False.
2. yes, the model is in eval model both the times (adapter and merge)
3. Parameters : temperature=0, max_new_tokens=100, do_sample=False, repetition_penalty=1.1
4. This is the input prompt:
   [INST] Complete the JAVA code below. Please wrap your code answer using ```:
@PutMapping("/updateProduct")
      public ResponseEntity<?> updateProduct(@RequestBody Product product) {

[/INST]
5. While i checked the logits, when i provide max_new_tokens are 100, first 52 tokens are same and rest of them are different.

Merged model output : tensor([[    1,   518, 25580, 29962, 25034,   278,   435, 26612,   775,  2400,
        29889,  3529, 12244,   596,   775,  1234,   773,  4954,  6998,    13,
        29992, 22908, 15845, 11974,  5504,  7566,  1159,    13,    12,  3597,
        13291,  6691,  8169, 29958,  2767,  7566, 10394,  3089,  8434, 10969,
        3234, 29897,   426,    13,    13, 29961, 29914, 25580, 29962, 29871,
        7521,    13,  _**2202,   426,    13,    12,    12,    12,   361,   313,
        4704,  2804,  1870,  2607,  3234, 29889,   657,  1204,   580,  2804,
        1870, 29897,   426,    13,    12,    12,    12,    12,  4704,  3170,
        29889,  7620, 29898,  4704,   416,    13,    12,    12,    12, 29913,
        1683,   426,    13,    12,    12,    12,    12,  2457,   716, 13291,
        6691, 25299,   703,  7566,   338,   451,  2854,   613,  9056,  5709,
        29889, 29933,  3035, 29918, 16244,   416,    13,    12,    12,    12,
        29913,    13,    12,    12, 29913,  4380,   313,  2451,   321, 29897,
        426,    13,    12,    12,    12, 12177, 29889,  2158,  7264, 16190,
        29898, 29872,   416,    13,    12,    12,    12, 14480, 17070**_]],
    device='cuda:0')


Adapter model ouput: tensor([[    1,   518, 25580, 29962, 25034,   278,   435, 26612,   775,  2400,
         29889,  3529, 12244,   596,   775,  1234,   773,  4954,  6998,    13,
         29992, 22908, 15845, 11974,  5504,  7566,  1159,    13,    12,  3597,
         13291,  6691,  8169, 29958,  2767,  7566, 10394,  3089,  8434, 10969,
          3234, 29897,   426,    13,    13, 29961, 29914, 25580, 29962, 29871,
          7521,    13, **_29992, 22908, 15845, 11974,  5504,  7566,  1159,    13,
            12,  3597, 13291,  6691,  8169, 29958,  2767,  7566, 10394,  3089,
          8434, 10969,  3234, 29897,   426,    13,    12,    12,   458, 10318,
           278,  3234,   297,   278,  2566,    13,    12,    12,  4704,  3170,
         29889,  5504,  7566, 29898,  4704,   416,    13,    12,    12,  2457,
           716, 13291,  6691, 29966,  5961,  5506,  5709, 29889,  8949,   416,
            13,    12, 29913,    13, 28956,    13,  4013,   775, 17645,   263,
         16759,  1319,  3450, 16248,   393, 11217,   263,  3234,   297,   278,
          2566,   746,   372, 20586,   263,   349,  2692,  2009,   411,   263,
          4663, 20092,  6943,   278,  4784,  3234,  2472, 29889,   450_**]],
       device='cuda:0')

**Comment by BenjaminBossan** [comment]
Thanks for investigating further. What I suspect is happening is that due to small rounding errors, you start to find deviations in longer sequences. Especially at lower floating point precisions, some deviations are unfortunately expected (with floating points, `(a + b) + c == a + (b + c)`, `a * (b + c) == a * b + a * c`, etc. are not necessarily true).

Again, if you compare the predicted logits, you'll most likely find that they're extremely close, but not 100% identical.

**Comment by akhilamartia** [comment]
> Thanks for investigating further. What I suspect is happening is that due to small rounding errors, you start to find deviations in longer sequences. Especially at lower floating point precisions, some deviations are unfortunately expected (with floating points, `(a + b) + c == a + (b + c)`, `a * (b + c) == a * b + a * c`, etc. are not necessarily true).
> 
> Again, if you compare the predicted logits, you'll most likely find that they're extremely close, but not 100% identical.

1. Okay the model returns the input along with the output, so the first 52 tokens are input getting repeated in both cases hence they are exactly the same and the rest of the output tokens are quite different.
2. From the text we see that merged model gives out more contextual completion and adapter code isn't, the relevance to the fine tuned data is nearer when merged model is used. while adapter provides more generic code completion. I cannot post client internal code to prove it, sorry.
3. Does small rounding errors would make it so different?
4. We are using the base model for a classification problem and finetuned model for code completion, having 2 models up was not very resource effective hence we went to the adapter approach, is there a way to make the adapter behave more like the merged model if not 100%. 

output merged 100 tokens: try {
			if (product != null && product.getId() != null) {
				productService.save(product);
			} else {
				return new ResponseEntity<>("Product is not valid", HttpStatus.BAD_REQUEST);
			}
		} catch (Exception e) {
			Utils.printStackMsg(e);
			LOGGER 

output adapter 100 tokens: // Update the product in the database
		productService.updateProduct(product);
		return new ResponseEntity<>(HttpStatus.OK);
	}

**Comment by akhilamartia** [comment]
> To add what @BenjaminBossan we also have strong CI tests that makes sure the logits are the same when the adapter is merged. Make also sure to always use the greedy generation strategy `do_sample=False` when calling `generate`. A full reproducer would be really helpful !

I have updated the merged code and then again please check my replies to Benjamin. I did try the do_sample and it didnt change.

**Comment by BenjaminBossan** [comment]
> 1. Okay the model returns the input along with the output, so the first 52 tokens are input getting repeated in both cases hence they are exactly the same and the rest of the output tokens are quite different.

Ah I see, thanks.

> * From the text we see that merged model gives out more contextual completion and adapter code isn't, the relevance to the fine tuned data is nearer when merged model is used. while adapter provides more generic code completion. I cannot post client internal code to prove it, sorry.
> * Does small rounding errors would make it so different?

Hmm, this is very tough to answer in abstract. Generally, I would think the differences should not be that big. Did you run multiple examples and always see the same pattern or was it just for one specific example?

IIUC, you say that the merged model gives better results but you would prefer not to merge it so that you can keep using the base model without the adapter. What I find strange is that if anything, I would expect the merged model to be worse. The reason for that is that during training, the model is also not merged, so the unmerged state should reflect training the best.

Again, if you could generate the logits (not the tokens) for merged vs unmerged and compare them, that could give us more insights.

**Comment by zysNLP** [comment]
@BenjaminBossan 
Hi, Thank you for your great work! I encountered the same problem. 
After I used the qwen model and the conversation data to fine-tune, the lora parameters were merged and did not match the do_predict method. In fact, our own requirements are not high, we only need to merge the predicted results after lora parameters, and do_predict (model validation) results are consistent at each time. Please note that do_predict should exactly the same with merged, after all, the stage of model verification is also the stage of manual verification, so naturally we want to merge the code will be completely consistent. 
I have two questions: 
1. Have you implemented the method of do_predict that I mentioned before and the inference result of integrating lora parameters is consistent? 
2. I think this should be a basic work in our project, but sorry to not find it util now.

I show some codes below:

Firstly，模型do_predict的验证脚本（多次验证结果相同，稳定性很好）：
CUDA_VISIBLE_DEVICES=0 python src/train_bash.py \
    --stage sft \
    --do_predict \
    --model_name_or_path /root/autodl-tmp/common_models/qwen1.5-14b-chat \
    --adapter_name_or_path /path/to/LLaMA-Factory/output/qwen1.5-14b-chat/checkpoint-2000 \
    --dataset email_valid \
    --template default \
    --finetuning_type lora \
    --output_dir ./output/qwen1.5-14b-chat_predict_2000 \
    --per_device_eval_batch_size 4 \
    --predict_with_generate \
    --quantization_bit 4

Secondly ，After merged Big different融合后的数据进行推理的脚本，结果差异巨大

```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig

class Predictor:
    def __init__(self, model_path, prompt=""):
        self.device = torch.device("cuda")  # the device to load the model onto
        self.model_path = model_path
        self.prompt = prompt

    def init_model(self, pred_by_int4=True):
        tokenizer = AutoTokenizer.from_pretrained(self.model_path, trust_remote_code=True)
        if pred_by_int4:
            quantization_config = BitsAndBytesConfig(
                load_in_4bit=True, bnb_4bit_compute_dtype=torch.float16)
            model = AutoModelForCausalLM.from_pretrained(
                self.model_path, quantization_config=quantization_config, trust_remote_code=True)
        else:
            model = AutoModelForCausalLM.from_pretrained(
                self.model_path, trust_remote_code=True)

        return tokenizer, model

    def predict(self, input_text):

        print("Starting init model...")
        tokenizer, model = self.init_model(pred_by_int4=True)
        print("Starting predict...")
        messages = [
            {"role": "system", "content": ""},
            {"role": "user", "content": input_text}
        ]

        text = tokenizer.apply_chat_template(
            messages,
            tokenize=False,
            add_generation_prompt=True
        )
        model_inputs = tokenizer([text], return_tensors="pt").to(self.device)

        generated_ids = model.generate(
            model_inputs.input_ids,
            max_new_tokens=512
        )
        generated_ids = [
            output_ids[len(input_ids):] for input_ids, output_ids in zip(model_inputs.input_ids, generated_ids)
        ]

        response = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)[0]
        print(f"{input_text = }")
        print(f"{response = }")
        return response

if __name__ == "__main__":
    predictor = Predictor(model_path="/path/to/LLaMA-Factory/output/qwen1.5-14b-chat_fusion")
    predictor.predict(input_text="'Email content: My order did not arrive. How can I get my refund?. Please reply this email using a template'")

```
3. results：第二部分的结果：

```shell
(ys_310) root@autodl-container-24e51191ae-913f8fd1:~/autodl-tmp/my/LLaMA-Factory# python test_fusion.py 
Starting init model...
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
`low_cpu_mem_usage` was None, now set to True since model is quantized.
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████| 15/15 [00:11<00:00,  1.28it/s]
Starting predict...
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:151645 for open-end generation.
input_text = "'Email content: My order did not arrive. How can I get my refund?. Please reply this email using a template'"
response = 'Dear <{$order.ship_first_name}> <{$order.ship_last_name}>, Thanks for your mail. If you have any problem with your order, please contact us as soon as possible. Generally, it will be 15-25 working days for orders to be delivered overseas. So we kindly request you to wait 15-20 worki'
(ys_310) root@autodl-container-24e51191ae-913f8fd1:~/autodl-tmp/my/LLaMA-Factory# 
(ys_310) root@autodl-container-24e51191ae-913f8fd1:~/autodl-tmp/my/LLaMA-Factory# python test_fusion.py 
Starting init model...
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
`low_cpu_mem_usage` was None, now set to True since model is quantized.
Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████| 15/15 [00:10<00:00,  1.40it/s]
Starting predict...
The attention mask and the pad token id were not set. As a consequence, you may observe unexpected behavior. Please pass your input's `attention_mask` to obtain reliable results.
Setting `pad_token_id` to `eos_token_id`:151645 for open-end generation.
input_text = "'Email content: My order did not arrive. How can I get my refund?. Please reply this email using a template'"
response = 'Dear <{$order.ship_first_name}> <{$order.ship_last_name}>,\nThanks for your email. Please be informed that your order has been shipped out on <{$order.date_ship}> to the following address:\n<{$order.ship_first_name}> <{$order.ship_last_name}>\n<{$order.ship_street1}>\n<{$order.ship_street2}>\n<{$order.ship_city}>\n<{$order.ship_state}>\n<{$order.ship_zip}>\n<{$order.ship_country}>\nThe tracking number is <{$order.follow_number}>. The shipper is <{$order.carrier}>.\nAccording to logistics information, your order has already been delivered.\nIf you have any questions, please feel free to contact us. Thank you!'
```

for easy, show two repsonse below：为了方便，把两次运行的response分别打印到底下：
response = 'Dear <{$order.ship_first_name}> <{$order.ship_last_name}>, Thanks for your mail. If you have any problem with your order, please contact us as soon as possible. Generally, it will be 15-25 working days for orders to be delivered overseas. So we kindly request you to wait 15-20 worki'

response = 'Dear <{$order.ship_first_name}> <{$order.ship_last_name}>,\nThanks for your email. Please be informed that your order has been shipped out on <{$order.date_ship}> to the following address:\n<{$order.ship_first_name}> <{$order.ship_last_name}>\n<{$order.ship_street1}>\n<{$order.ship_street2}>\n<{$order.ship_city}>\n<{$order.ship_state}>\n<{$order.ship_zip}>\n<{$order.ship_country}>\nThe tracking number is <{$order.follow_number}>. The shipper is <{$order.carrier}>.\nAccording to logistics information, your order has already been delivered.\nIf you have any questions, please feel free to contact us. Thank you!'

**Comment by AdamCassidy** [comment]
I have a similar issue where model_8bit.add_adapter(LoraConfig(etc.)) with model_8bit.enable_adapters() causes a "peft not installed" error for some reason for codellama/CodeLlama-7b-hf and meta-llama/Meta-Llama-3-8B doesn't bring up that error.
